feat: Update cloud-init customization

dlipovetsky opened the pull request:
## Description
Changes relative to upstream:
* Add explanatory comments
* Do not use stderr output of preKubeadmCommands indicate an error with bootstrapping

Changes relative to our fork:
* Do not enable IPv6
* Do not remove cloud-init logs and seed
* Do not disable VMware customization
* Do not disable network configuration
* Do not truncate cloud-init-output.log
* Do not report status of HTTP proxy configuration
* Do not configure cloud-init to remove SSH keys on first boot
* Remove commands that are already executed as a result of being defined in `preKubeadmCommands`

dlipovetsky committed:
feat: Update cloud-init customization

Changes relative to upstream:
* Add explanatory comments
* Do not use stderr output of preKubeadmCommands indicate an error with
  bootstrapping

Changes relative to our fork:
* Do not enable IPv6
* Do not remove cloud-init logs and seed
* Do not disable VMware customization
* Do not disable network configuration
* Do not truncate cloud-init-output.log
* Do not report status of HTTP proxy configuration
* Do not configure cloud-init to remove SSH keys on first boot
* Remove commands that are already executed as a result of being defined in `preKubeadmCommands`

supershal commented on controllers/cluster_scripts/cloud_init.tmpl:
```diff
@@ -33,21 +43,21 @@ write_files:
-    hostname "{{ .MachineName }}"
-    echo "::1         ipv6-localhost ipv6-loopback" >/etc/hosts
-    echo "127.0.0.1   localhost" >>/etc/hosts
-    echo "{{ .MachineName }}" >/etc/hostname
```
we should keep this entry ` echo "{{ .MachineName }}" >/etc/hostname`

dlipovetsky replied:
Interesting; even without this entry, in my development cluster, `/etc/hostname` contains the machine name :thinking:

dlipovetsky replied:
I think cloud-init sets `/etc/hostname`, based on this configuration:
https://github.com/mesosphere/cluster-api-provider-cloud-director/blob/66ecf82670254cafae0e84f9fc30d6fa454e5e28/controllers/cluster_scripts/cloud_init.tmpl#L83-L84

supershal replied:
Thanks. That makes sense.

supershal commented on controllers/cluster_scripts/cloud_init.tmpl:
```diff
@@ -58,12 +68,8 @@ write_files:
-- 'sudo cat /dev/null > /var/log/cloud-init-output.log'
+- 'cloud-init clean'
 {{ if .ControlPlane }}
-- '[ ! -f /run/kubeadm/konvoy-set-kube-proxy-configuration.sh] && sudo reboot'
```
were you able to boot the VM and create cluster after removing this file? I remember that pre kubeadm commands were failing If I removed them. I will test it out later to confirm

dlipovetsky replied:
Good question!

You're right that any preKubeadmCommand that requires an ordinary file in `/run` will fail after a reboot, because ordinary files in `/run` do not persist across reboots. We recently (in https://github.com/mesosphere/konvoy2/pull/2337) moved all patch scripts from `/run` to `/etc` for this reason.

Your question made me wonder about the two `reboot` calls left in this template:
https://github.com/mesosphere/cluster-api-provider-cloud-director/blob/66ecf82670254cafae0e84f9fc30d6fa454e5e28/controllers/cluster_scripts/cloud_init.tmpl#L72-L80

It seems harmless to reboot if the kubeadm config (`/run/kubeadm/kubeadm.yaml` or `/run/kubeadm/kubeadm-join-config.yaml`) is not not there (yet?).
 
But if we reboot because the bootstrap script (`/root/control_plane.sh` or `/root/node.sh`) is missing, and the kubeadm config happens to already be present, we will lose the kubeadm config after the reboot, leading to further reboots, without end.

At this time (https://github.com/mesosphere/cluster-api-provider-cloud-director/commit/66ecf82670254cafae0e84f9fc30d6fa454e5e28), I can successfully reboot either a control plane, or worker machine.

I think it may be better to remove the remaining `reboot` calls. I will experiment.

dlipovetsky replied:
I've added a comment that explains why the reboot call is necessary. I've also moved these checks out to their own script, and use a separate log file to keep track.

This is what the log looks like:
```log
# cat /var/log/capvcd/replace-userdata-files.log
2023-10-17 22:07:37 Checking for kubeadm configuration file
2023-10-17 22:07:37 kubeadm configuration file not found, cleaning cloud-init cache and rebooting
2023-10-17 22:08:12 Checking for kubeadm configuration file
2023-10-17 22:08:12 kubeadm configuration file found, exiting
```

supershal replied:
Pretty clever. just iterating the logic: until the user-data is available, the vm will reboot and cloud-init will run it as if its first boot. once it is user-data are available, the bootstrap.sh will run kubeadm init/join.

dlipovetsky replied:
> until the user-data is available, the vm will reboot and cloud-init will run it as if its first boot. once it is user-data are available, the bootstrap.sh will run kubeadm init/join.

Correct. The upstream cloud-init also did this, but used in-line commands, instead of a script, and the reason behind everything wasn't given.

dlipovetsky committed:
Simplify cloud init

* Use shell script to clean cloud-init cache and reboot.
* Fix error handling of bootstrap script. Do not interpret stderr output
  as an indicator of failure. Do not rely on trap and errexit, because
  it does not work for command lists.
* Include last lines of output for error context.
* Ensure we have an IPv4 address for localhost.
* Remove unnecessary cloud-init configuration to preserve SSH host keys.

supershal reviewed (approved):
Thank you for testing this out.

dkoshkin reviewed (approved):
Thank you for all the comments and really digging deep into this! Great changes and so much simpler to understand.
